src/System.Text.Json/tests/Utf8JsonWriterTests.cs

```diff
@@ -580,6 +581,71 @@ public async Task UseAfterDisposeInvalidAsync(bool formatted, bool skipValidatio
             Assert.Throws<ObjectDisposedException>(() => jsonUtf8.Reset(output));
         }
 
+        [Fact]
+        public void FlushToStreamThrows_RemainsInConsistentState()
+        {
+            var stream = new ThrowingFromWriteMemoryStream();
+            var jsonUtf8 = new Utf8JsonWriter(stream);
+
+            // Write and flush some of an object.
+            jsonUtf8.WriteStartObject();
+            jsonUtf8.WriteString("someProp1", "someValue1");
+            jsonUtf8.Flush();
+
+            // Write some more, but fail while flushing to write to the underlying stream.
+            stream.ExceptionToThrow = new FormatException("uh oh");
+            jsonUtf8.WriteString("someProp2", "someValue2");
+            Assert.Same(stream.ExceptionToThrow, Assert.Throws<FormatException>(() => jsonUtf8.Flush()));
+
+            // Write some more.
+            stream.ExceptionToThrow = null;
+            jsonUtf8.WriteEndObject();
+
+            // Disposing should not fail.
+            jsonUtf8.Dispose();
+        }
+
+        [Fact]
+        public async Task FlushAsyncToStreamThrows_RemainsInConsistentState()
+        {
+            var stream = new ThrowingFromWriteMemoryStream();
+            var jsonUtf8 = new Utf8JsonWriter(stream);
+
+            // Write and flush some of an object.
+            jsonUtf8.WriteStartObject();
+            jsonUtf8.WriteString("someProp1", "someValue1");
+            await jsonUtf8.FlushAsync();
+
+            // Write some more, but fail while flushing to write to the underlying stream.
+            stream.ExceptionToThrow = new FormatException("uh oh");
+            jsonUtf8.WriteString("someProp2", "someValue2");
+            Assert.Same(stream.ExceptionToThrow, await Assert.ThrowsAsync<FormatException>(() => jsonUtf8.FlushAsync()));
+
+            // Write some more.
+            stream.ExceptionToThrow = null;
+            jsonUtf8.WriteEndObject();
+
+            // Disposing should not fail.
+            await jsonUtf8.DisposeAsync();
+        }
```

Comment:
I assume the idea here is that by the call to Flush failing that the data is still held in the writer and gets eventually pushed again.  Checking that the Stream has the expected contents in the end seems good in that case.

Reply:
I actually don't think that part matters too much; what matters more is that Dispose{Async} doesn't fail / throw because of the previous failure.  Currently what's happening is Dispose{Async} are throwing due to the corruption caused by the previous stream call throwing.  But, I can certainly add a check for that if you think it's useful; my assumption has been that if data failed to be written to the stream such that an exception is propagated from that stream write/flush, the data in it doesn't actually matter.

Reply:
I guess it's just not clear to me what corruption here is worth being fixed and what corruption isn't (and then beyond what was "worth" fixing, what "is" fixed/prevented).

If the actual data is being written off in the throw case, that's... fine... I guess.  But if the changes mean that now the writer keeps the data buffered until the next call to Flush(Async) that seems worth testing so it doesn't regress.

Reply:
> I guess it's just not clear to me what corruption here is worth being fixed and what corruption isn't

See the issue being fixed by this. Without the fix, you get an exception "Cannot advance past the end of the buffer, which has a size of 256" (and an assert in debug).  With the fix, you get an exception "Synchronous operations are disallowed."